### DISCUSSION TITLE
Looking for a script to backup/restore MUxx-Datasets only

### DISCUSSION THREAD

**Discussion by Jay-T88** [Mr-MIBonk/M.I.B._More-Incredible-Bash#205] [General]
**1st let me thank you for your amazing work - highly appreciated!!!**

**Question: Is there an easy way to just back&restore sound datasets only: MU1467-datasets.csv**

I'm adjusting my sound set according to the great work of Alex https://mqb-blog.com/ and I'd like to write back the modifications without doing a whole restore of MMI Unit.

I was able to read/write 2x EQ sound sets by using Car Scanner Pro, but unfortunately 0x3900 is missing so far. I've extracted the according data set from MU1467-datasets.csv, adjusted them with templates from Alex and injected them back to MU1467-datasets.csv. Now I'd like to write back just MU1467-datasets.csv only.
- 0x3000 - audio parameter Sound Details -> works 
- 0x3B00 - audio parameter Sound Configuration -> works
- 0x3900 - audio parameter Sound for Open Roof / Cabrio  -> missing in Car Scanner Pro

Thanks for any ideas and help....

**Comment by harman-f** [comment]
M.I.B has a built in patch function for these sound datasets and will also create a backup of these ;)

**Reply by Jay-T88** [reply]
Thanks I know that Backup command backs up everything.. but how can I just back/restore datasets?
Is it this menu point? But datasets are in CSV format - getting confused with ".raw"

![image](https://github.com/Mr-MIBonk/M.I.B._More-Incredible-Bash/assets/132943993/66fc7e75-94ff-4075-bdaa-d5cd31e9e95f)

**Reply by harman-f** [reply]
Datasets csv hold many, also some onknown datasets. 

Storage .raw holds some, most important datasets, which can be restored by copying the file back from backup. 

There is no automated restore function for the csv file. (Yet)

**Comment by Jay-T88** [comment]
I just checked MU1467-storage1.raw + 2 -> content is very different to MU1467-datasets.csv.
File sizes are also very different 19kb+19kb vs. 75kb
So only chance is to modify MU1467-datasets.csv to my needs and do a full restore than...

**Reply by harman-f** [reply]
Two different things. 
csv contains a list of dumped datasets. 

Storage .raw is a file generated by unit to store certain datasets.

**Comment by Jay-T88** [comment]
Any chance to run an individual script to read & write MU1467-datasets.csv only?

![M I B  Menu Run individual Script](https://github.com/Mr-MIBonk/M.I.B._More-Incredible-Bash/assets/132943993/1fc7dfd9-4238-4183-8bb8-c74e38be4717)

**Reply by harman-f** [reply]
Of course, this would be possible.
Write a script to parse the csv and write datasets to persistence.

But why not start with storage .raw?

**Comment by Jay-T88** [comment]
Thanks to Alex https://github.com/NumberOneBot I was able to write the dataset by using his documentation for a script to 0xC0030018 = 0x003900 -> Audio Sound Param. EQ Details open roof / convertible.
Seems to work just need a patch for my version MU1467 to fix SVM or update to supported Firmware MU1479/MU1489.

**How to flash dataset without ODIS/VCP**
1. create file named command.sh and copy following content into /mod folder on M.I.B sd card
2. go to GEM "Run individual script" that will execute command.sh following example will write HMI dataset to 0x2F00
---
export PATH=:/proc/boot:/sbin:/bin:/usr/bin:/usr/sbin:/net/mmx/bin:/net/mmx/usr/bin:/net/mmx/usr/sbin:/net/mmx/sbin:/net/mmx/mnt/app/armle/bin:/net/mmx/mnt/app/armle/sbin:/net/mmx/mnt/app/armle/usr/bin:/net/mmx/mnt/app/armle/usr/sbin
export LD_LIBRARY_PATH=/net/mmx/mnt/app/root/lib-target:/net/mmx/mnt/eso/lib:/net/mmx/eso/lib:/net/mmx/mnt/app/usr/lib:/net/mmx/mnt/app/armle/lib:/net/mmx/mnt/app/armle/lib/dll:/net/mmx/mnt/app/armle/usr/lib
export IPL_CONFIG_DIR=/etc/eso/production

#write eco hmi dataset
on -f mmx /eso/bin/apps/pc b:0:0xC0030016 0a0a010a05010a03010a04000a0a0a0501010909030909030909030a01010a0821537b0a0a123221537b0a0503010a0503011114160507090204061014180e101204060814121e28041e030a1eb4031e0521501e0f0a0a000000000000000000000032465cd4

#store changes to persistence
on -f mmx /eso/bin/apps/pc b:0:1 0

#reboot to save settings to unit!
on -f rcc /usr/apps/mib2_ioc_flash reboot

**Reply by harman-f** [reply]
https://mibwiki.one/doc/persistence-datasets-4uXMspDHAc

**Reply by snuew** [reply]
Does the MIB unit need to be patched to be able run the script?

**Comment by Muilperen** [comment]
When I try to run the custom script, I get below error and then the unit reboots immediately.
![screen](https://github.com/Mr-MIBonk/M.I.B._More-Incredible-Bash/assets/134429730/dfa6d6bc-9134-47dd-a509-a4cf1344f590)


Below is the script I'm trying to run. Perhaps anyone has an idea why it's not working?

```
export PATH=:/proc/boot:/sbin:/bin:/usr/bin:/usr/sbin:/net/mmx/bin:/net/mmx/usr/bin:/net/mmx/usr/sbin:/net/mmx/sbin:/net/mmx/mnt/app/armle/bin:/net/mmx/mnt/app/armle/sbin:/net/mmx/mnt/app/armle/usr/bin:/net/mmx/mnt/app/armle/usr/sbin
export LD_LIBRARY_PATH=/net/mmx/mnt/app/root/lib-target:/net/mmx/mnt/eso/lib:/net/mmx/eso/lib:/net/mmx/mnt/app/usr/lib:/net/mmx/mnt/app/armle/lib:/net/mmx/mnt/app/armle/lib/dll:/net/mmx/mnt/app/armle/usr/lib
export IPL_CONFIG_DIR=/etc/eso/production

#write eco hmi dataset
on -f mmx /eso/bin/apps/pc b:0:0xC0030014 20060e0050006f000700fa005a000700fa005a000700fa005a010e183800785a0405003c0000000000000000000000000000000000000006071f40000000000000000000000000000000000000000a0c0d0f1112141617191b1c1e2021232526280a0c0e10131517191b1e20222426282b2d2f310a0d101215181b1d202326292b2e313436393c0a0e1115181c1f23262a2d3134383b3f4246490a0e12171b1f23282c3034383d4145494e52560a0f14191e23282d32373c41464b50555a5f640a10151b21272c32383e43494f545a60666b71060a00960000000000000000000000000000000000000000c8012c0a141e1e1e1e1e282828323232323c3c3c3c3c000300361900000000000000000000000000000000000a04efd700000000000000000000000000000000000a06ceec00000000000000000000000000000000000d008bdc000a002a00001a08f3d4000a25e4f9000000000a00320000000000000000000000000000000000001100a61400000000000000000000000000000000001807291a0000000000000000000000000000000000150405e0001702e2d9000a2710f9000d012ce1000000000700320000000000000000000000000000000000001100a61400000000000000000000000000000000001807291a0000000000000000000000000000000000150405e0001702e2d9000a2710f9000d012ce1000000000300361900000000000000000000000000000000000a04efd700000000000000000000000000000000000a06ceec00000000000000000000000000000000000d008bdc0007002a00001a08f3d4000725e4f9000000060703e80000000000000000000000000000000000060703e80000000000000000000000000000000000060703e80000000000000000000000000000000000060703e800060703e800060703e800060703e800000000060703e80000000000000000000000000000000000060703e80000000000000000000000000000000000060703e80000000000000000000000000000000000060703e800060703e800060703e800060703e800000000060703e80000000000000000000000000000000000060703e80000000000000000000000000000000000060703e80000000000000000000000000000000000060703e800060703e800060703e800060703e800000000060703e80000000000000000000000000000000000060703e80000000000000000000000000000000000060703e80000000000000000000000000000000000060703e800060703e800060703e800060703e800000000060703e80000000000000000000000000000000000060703e80000000000000000000000000000000000060703e80000000000000000000000000000000000060703e800060703e800060703e800060703e800000000060703e80000000000000000000000000000000000060703e80000000000000000000000000000000000060703e80000000000000000000000000000000000060703e800060703e800060703e800060703e800000000060703e80000000000000000000000000000000000060703e80000000000000000000000000000000000060703e80000000000000000000000000000000000060703e800060703e800060703e800060703e800000000060703e80000000000000000000000000000000000060703e80000000000000000000000000000000000060703e80000000000000000000000000000000000060703e800060703e800060703e800060703e8000000000000000000000a1400000000140000000014000000001400000000320032000a320032000a4d35692a

#store changes to persistence
on -f mmx /eso/bin/apps/pc b:0:1 0

#reboot to save settings to unit!
on -f rcc /usr/apps/mib2_ioc_flash reboot
```

**Reply by harman-f** [reply]
did you check to only use unix EOL?

Please upload the command.sh

**Reply by Muilperen** [reply]
As someone who is not a programmer, this did not appear to me :)
Thank you so much for pointing me in the right direction. With your help I was able to get the script running with no issue.
I really appreciate the effort, keep up the good work!

**Reply by Jet-ZJ** [reply]
Seems your last 2 bytes CRC (692a) is not correct. Is it working fine for you?
Maybe I'm woring.

**Comment by traviesojaen83** [comment]
Hello everyone, first I thank you for your work, but I need help from one of you, I need to be able to restore my mib unit of an a4 b9, I have problems with the pre sense system of the vehicle after patching. I need to restore the drive to before the patch and uninstall more incredible bash. Thanks in advance

**Comment by Tomt97** [comment]
Hi all, @Jay-T88 @harman-f , I hope you guys can help., 
TLDR: replace my original unit(internal amp) with a S5 head unit(external amp).    I was able to switch it to internal amp, but the sound just not right., sound super flat.   Unit is a MHI2Q_US

been looking around and trying to learn this thing but just not getting the full understand.   I'm add
[MU1316-datasets.csv](https://github.com/user-attachments/files/16855977/MU1316-datasets.csv) file from MIB from my original unit.    I would like to copy this sound setting over to the new S5 head unit

per Alex file,  does everyone use the same code in the command.sh?

```
export PATH=:/proc/boot:/sbin:/bin:/usr/bin:/usr/sbin:/net/mmx/bin:/net/mmx/usr/bin:/net/mmx/usr/sbin:/net/mmx/sbin:/net/mmx/mnt/app/armle/bin:/net/mmx/mnt/app/armle/sbin:/net/mmx/mnt/app/armle/usr/bin:/net/mmx/mnt/app/armle/usr/sbin
export LD_LIBRARY_PATH=/net/mmx/mnt/app/root/lib-target:/net/mmx/mnt/eso/lib:/net/mmx/eso/lib:/net/mmx/mnt/app/usr/lib:/net/mmx/mnt/app/armle/lib:/net/mmx/mnt/app/armle/lib/dll:/net/mmx/mnt/app/armle/usr/lib
export IPL_CONFIG_DIR=/etc/eso/production

#write eco hmi dataset
on -f mmx /eso/bin/apps/pc b:0:0xC0030016 0a0a010a05010a03010a04000a0a0a0501010909030909030909030a01010a0821537b0a0a123221537b0a0503010a0503011114160507090204061014180e101204060814121e28041e030a1eb4031e0521501e0f0a0a000000000000000000000032465cd4

#store changes to persistence
on -f mmx /eso/bin/apps/pc b:0:1 0

#reboot to save settings to unit!
on -f rcc /usr/apps/mib2_ioc_flash reboot
```

**Comment by Tomt97** [comment]
Also now sure if this help any?  But there is what MIB Show.  I want to get the data to be the same as the old unit onto the new 
![IMG_0891](https://github.com/user-attachments/assets/76ccf703-df49-4b7f-86b5-b70e55113f11)
![IMG_0895](https://github.com/user-attachments/assets/723a3113-3fce-4442-92fc-e24a6f7d4919)

**Comment by Tomt97** [comment]
Anyone know how to if we can update just the mib dataset.csv file?  BUMP
